lib/user_communication_defender.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+# There are various ways within Discourse that a user can prevent
+# other users communicating with them. The purpose of this class is to
+# find which of the target users are ignoring, muting, or preventing
+# private messages from the acting user, so we can take alternative
+# action (such as raising an error or showing a helpful message) if so.
+class UserCommunicationDefender
```

Comment:
Design thought: Would like some general feedback about the naming and structure of this class. I like that it is very self contained and will be portable for plugins.

Reply:
Some ideas:

* `InterUserScreener`
* `InterUserCommunicationScreener`
* `UserCommunicationScreener`
* `UserCommsScreener`? 

I'm not really sure about `Defender` because I think about `Attack` when I see that. Not sure if any of the actions we're taking between users really fall into the `attack` bucket.

Reply:
I like `UserCommsScreener` I think, that is shorter, and you are right about the attack part of it. They aren't attacking them (hopefully) just trying to mention them or send them a PM

Reply:
Maybe just `UserCommScreener` without the small `s`? I think it'll be less prone to typos.

Reply:
Yeah good idea